certificate-test-fixture/src/main/java/io/netty/testcert/X509Bundle.java

```diff
@@ -0,0 +1,345 @@
+     * @param keyPair The key pair.
+     */
+    public static X509Bundle fromCertificatePath(
+            X509Certificate[] certPath, X509Certificate root, KeyPair keyPair) {
+        return new X509Bundle(certPath, root, keyPair);
+    }
+
+    /**
+     * Get the leaf certificate of the bundle.
+     * If this bundle is for a certificate authority, then this return the same as {@link #getRootCertificate()}.
+     * @return The leaf certificate.
+     */
+    public X509Certificate getCertificate() {
+        return certPath[0];
+    }
+
+    /**
+     * Get the PEM encoded string of the {@linkplain #getCertificate() leaf certificate}.
+     * @return The certificate PEM string.
+     */
+    public String getCertificatePEM() {
+        return toCertPem(certPath[0]);
+    }
+
+    /**
+     * Get the certificate path, starting with the leaf certificate up to but excluding the root certificate.
+     * @return The certificate path.
+     */
+    public X509Certificate[] getCertificatePath() {
+        return certPath.clone();
+    }
+
+    /**
+     * Get the certificate path, starting with the leaf certificate up to and including the root certificate.
+     * @return The certificate path, including the root certificate.
+     */
+    public X509Certificate[] getCertificatePathWithRoot() {
+        X509Certificate[] path = Arrays.copyOf(certPath, certPath.length + 1);
+        path[path.length - 1] = root;
+        return path;
+    }
+
+    /**
+     * Get the certificate path as a list, starting with the leaf certificate up to but excluding the root certificate.
+     * @return The certificate path list.
+     */
+    public List<X509Certificate> getCertificatePathList() {
+        return Arrays.asList(certPath);
+    }
+
+    /**
+     * Get the {@linkplain #getCertificatePath() certificate path} as a PEM encoded string.
+     * @return The PEM encoded certificate path.
+     */
+    public String getCertificatePathPEM() {
+        return toCertPem(certPath);
+    }
+
+    /**
+     * Get the key pair.
+     * @return The key pair.
+     */
+    public KeyPair getKeyPair() {
+        return keyPair;
+    }
+
+    /**
+     * Get the root certificate that anchors the certificate path.
+     * @return The root certificate.
+     */
+    public X509Certificate getRootCertificate() {
+        return root;
+    }
+
+    /**
+     * Get the {@linkplain #getRootCertificate() root certificate} as a PEM encoded string.
+     * @return The PEM encoded root certificate.
+     */
+    public String getRootCertificatePEM() {
+        return toCertPem(root);
+    }
+
+    private static String toCertPem(X509Certificate... certs) {
+        Base64.Encoder encoder = Base64.getMimeEncoder();
+        StringBuilder sb = new StringBuilder();
+        for (X509Certificate cert : certs) {
+            sb.append("-----BEGIN CERTIFICATE-----\r\n");
+            try {
+                sb.append(encoder.encodeToString(cert.getEncoded()));
```

Comment:
This will produce a long single-line certificate, not a truncated one, right?

Reply:
Yep. Most tools have some line wrapping for PEM certs, but there's no requirement so this is fine.

Reply:
It won't. The MIME encoder will wrap after col 76, which is in tune with whatever RFC it was that defined PEM.
Here's an example I grabbed from the debugger:

```
-----BEGIN CERTIFICATE-----
MIIBkTCCATegAwIBAgIUC88SLLFgbJ6Ax2Y4Uz4qMOn+qsgwCgYIKoZIzj0EAwIwIzEOMAwGA1UE
ChMFTmV0dHkxETAPBgNVBAMTCG5ldHR5LmlvMCIYDzIwMjQwOTAzMDIyMjEwWhgPMjAyNDA5MDUw
MjIyMTBaMBgxFjAUBgNVBAMTDWxlYWYubmV0dHkuaW8wWTATBgcqhkjOPQIBBggqhkjOPQMBBwNC
AASo0YfNrfQjxvVaD34VuGQbgMubAqgTDGdWcgBaQUq7XnqzkcAzvHO2zLeg2OsCU6e80xSdHemt
A3qtI98pJwxco1AwTjAWBgNVHSUBAQAEDDAKBggrBgEFBQcDAjA0BgNVHR8BAQAEKjAoMCagJKAi
hiBodHRwOi8vbG9jYWxob3N0OjYyMDA2L2NybC8xLmNybDAKBggqhkjOPQQDAgNIADBFAiEA6zO4
Rqx36kt4V0WNRi/6ef6be8RQ8g8Z+6/V6resx1wCIHY2El9m6J/yWHJwn/jUDgZhVk7LY01+oqRS
0FMtyq1y
-----END CERTIFICATE-----
```

Reply:
TIL. But it seems PEM specified 64 characters, not 76